### PR TITLE
Add test_bnd_prt_inflow with nascent `Inflow` class

### PR DESCRIPTION
**Pull request by JamesMcClung**
Eventually, `Inflow` will be moved and used by open particle boundary conditions. Its current state is mostly usable, but could be optimized by removing `std::function` calls.

### CODE REVIEW

File: src/libpsc/tests/test_bnd_prt_inflow.cxx [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+
+  template <typename Injectors>
+  void inject_into_boundary(Injectors& injectors_by_patch)
+  {
+    for (int patch_idx = 0; patch_idx < grid_.n_patches(); patch_idx++) {
+      const auto& patch = grid_.patches[patch_idx];
+
+      if (patch.off[inj_dim_idx_] != 0) {
+        continue;
+      }
```
Comment (germasch) [inline src/libpsc/tests/test_bnd_prt_inflow.cxx:186]: There is a `at_boundary_lo()` or something of that sort somewhere, but not sure it's easily reusable here.

File: src/libpsc/tests/test_bnd_prt_inflow.cxx [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+class TestInjector
+{
+public:
+  void operator()(psc::particle::Inject prt) { prts.push_back(prt); }
+
+  std::vector<psc::particle::Inject> prts;
+};
+
+double half() { return 0.5; }
```
Comment (germasch) [inline src/libpsc/tests/test_bnd_prt_inflow.cxx:209]: Minor nitpick: I'd prefer this to be static, though since it's just in a test source, I guess it doesn't really matter.

**Review by germasch — APPROVED** [review]
Looks good to me! (other than some minor nitpicks)

The one thing I don't love is the repetitiveness of the tests. But I'm not sure factoring out some common pieces won't make readability worse...

File: src/libpsc/tests/test_bnd_prt_inflow.cxx (hunk 1 above)
Comment (JamesMcClung) [reply src/libpsc/tests/test_bnd_prt_inflow.cxx]: I see `mrc_domain_at_boundary_lo`, but yeah, doesn't seem usable here

File: src/libpsc/tests/test_bnd_prt_inflow.cxx (hunk 2 above)
Comment (JamesMcClung) [reply src/libpsc/tests/test_bnd_prt_inflow.cxx]: Definitely `inj_dim_idx_` should have been static (and it now is), but did you mean also `half()`? It is too now.

Comment (germasch) [reply src/libpsc/tests/test_bnd_prt_inflow.cxx]: I'm less convinced that `inj_dim_idx_` should be static, I just meant the `half()`.

Comment (JamesMcClung) [reply src/libpsc/tests/test_bnd_prt_inflow.cxx]: If the injection dimension is y, and we care about pushing along the x and/or z dimensions as well, wouldn't that necessitate another check at each domain corner to ensure particles don't leave the x or z bounds? Or is it potentially even worse, because even when in bounds, they might end up getting injected into the wrong patch?

That currently can't happen (provided only a 1d `dim_...` is passed). Physically, I suppose it makes sense to support that for cases where the injected particle velocity distributions vary along the boundary, but that isn't the case for now at least. I also wouldn't want there to be some subtle pitfall for future users who might need such a feature, though. I'm personally inclined to decouple the `AdvanceParticle` dim and `inject_dim_idx`, and throw in an assert that they match with a comment explaining the problem.

More template args are definitely annoying, but maybe the performance benefits outweigh the ergonomic downsides; I don't know, and I'm happy to defer to your opinion on that.
